### ISSUE TITLE
No sound in PinballY at computer startup

### ISSUE THREAD

**Issue by drivin33** [mjrgh/PinballY#254]
I replaced the two sound cards in my cab PC and everything works well in Windows, VPX and all apps, except for PinballY. PinballY is set to automatically run when the computer boots. It loads fine but there is no sound when moving the wheel, startiung a game, etc,. But, if I select a game, play it, then PinballY returns to attract mode, suddenly all of the sounds now work as normal. I'm running the latest PinballY beta, latest release of Winidows 11.

**Comment by drivin33** [comment]
Thanks for your quick reply. I have a spare hard drive, so I did a fresh install of everything, VPX 10.8 RC 5, Windows 11, most recent versions of B2S server, Freezy DLLs, VPinMAME, etc. Did all of the Restributable C++ and DirectX updates. But the same problem with PinballY happens:

1. When PinballY is set to startup with Windows, no sound. But If I then start a table then exit back to PinballY, sound works.
2. When I set PinballY to start manually, sound works fine. In this scenario, I immediately run PinballY upon booting to the Windows desktop. 
3. One thing I just discovered - I record all the table videos with no sound. But there was one table that I inadvertently recorded with sound. As I was cycling thru the tables, the one table that I recorded the video with sound did play that sound. So it seems just the sounds made by cycling thru the tables with the flipper buttons that do not work. 

I have disabled all of the apps that startup automatically so it’s not that. So I am at a loss. One other thing - the issue that I mentioned in another bug report where PinballY crashes with some conflict with the PinDMD still happens with this new clean build. So that must be something with Freezy’s DLL.

**Comment by mjrgh** [comment]
I'm still thinking it's a startup timing thing, with PinballY starting too early, before DirectX is ready.  Did you try the delay option?  If you're using PinballY's options to configure the startup, just try setting the "After logon, pause before starting PinballY (seconds)" box to something fairly high, like 60 seconds (and if the first thing you try doesn't work, try something REALLY long, like 300 seconds).  Hopefully the real delay time needed is more like 10-20 seconds, but starting with something really long would be the quickest (ironically) way to confirm that some kind of delay would help, and once that's established, you find a minimum reliable delay experimentally by reducing it gradually until it stops working again.

If no amount of startup delay fixes it, then maybe the thing to do is to try a different WAY of starting it with the system.  For example, by using a Start Menu "startup" group shortcut instead of Task Scheduler.  Maybe it's the context under which you're currently launching it that's the problem, rather than just the timing.  But I'd rule out timing changes first since that seems more likely to me, given the symptoms you're describing.

**Comment by drivin33** [comment]
I followed your advice and that fixed it. A 10 second delay was what worked. Thanks! But having done that, I find that whatever process was causing this issue must be taking focus from PinballY as at boot up, I now need to have a mouse handy and click on PinballY to regain focus and be able to scroll through the tables. I’m aware of the focus option in settings, and I’ve tried that with various time delays but no luck. Any ideas?

**Comment by mjrgh** [comment]
Glad the startup delay worked for the sound problem. 

Focus issues are tough. Can you tell where focus lands? My guess is that it's going to the Windows desktop - you can test that by pressing cursor arrow keys and seeing if that affects which Windows desktop icon is selected. 

If focus is NOT in the desktop but in some other app, the fix would probably be to either not launch that app at startup, or tinker with the startup timing (either of PinballY or the other app, if it has any way to control it) to make sure that PinballY always starts after the other program. Windows generally gives focus to the last program launched.

If focus lands in the desktop, it's probably still some other program launch happening after PinballY, but in that case the other program is probably either minimizing/hiding its main window, or just terminating. In either case, focus will go back to the Windows desktop.  So the solution again is to try to ensure that PinballY launches after the other program.

Windows has about 5 different ways to launch programs at startup, and doesn't expose a lot of user controls over the launch order or timing, so there's not a good one-size-fits-all fix I know of.  About all I can recommend is the same process I suggested earlier for the sound problem, of making the PinballY startup delay time really super ridiculously impractically long, like 5 minutes, just to see if that resolves the focus problem.  If it does, you can then experiment with successively shorter times until it stops working, to establish the minimum required delay.  If no amount of delay works, I'd probably start tinkering with other launch-at-startup options to see if a different approach launches it and leaves focus in the right place.

**Comment by drivin33** [comment]
When PinballY starts after the 10 second delay, the taskbar icon flashes with an orange background which typically indicates that the app needs attention for some reason. But there are no error message or anything else that would require attention. So it’t in this state where it kind of has the focus, but not really. Pressing the arrow keys does nothing because of the flashing PinballY taskbar icon. When I then click on PinballY, the taskbar icon stops flashing and PinballY has complete focus. No amount of adjustment in PinballY focus delay or startup delay settings has made any difference. Every app or service that runs at startup was previously disabled, so it’s not that. I’ve tried putting PinballY in the Windows startup folder, but that made no difference.

**Comment by mjrgh** [comment]
Yeah, that's what Windows does (the taskbar icon flashing) when an application tries to bring itself to the foreground when something else is already in the foreground - Windows denies the focus change attempt and flashes the taskbar icon instead.  Windows is trying to be helpful, by preventing the background app from stealing focus when you're in the middle of typing into another program, but it makes it hard to get focus to go where you want it at times like this.  If you've ruled out the possibility that any other applications could be running at any point, then I guess the only thing left is that PinballY is getting launched without focus, which would be unusual - Windows usually brings a newly launched program into the foreground automatically at launch.  If it's not doing that, it must be some unusual launch condition, but without knowing the details of your system setup, I wouldn't know where to start looking. I guess if it were my system I'd try to reconstruct if I ever changed any of the default system settings with respect to the desktop, UAC, user accounts, or anything else that could affect program launch rules. Windows has so many weird special cases that it's hard to guess out of the blue.

**Comment by mjrgh** [comment]
The only other suggestion that comes to mind is what I was talking about earlier, of deleting the Task Scheduler startup entry (by disabling "start at login" in the PinballY options) and experiment with the various other launch-at-startup mechanisms that Windows offers.  There's the "Startup Apps" control panel, for example, and there's the %APPDATA%\Microsoft\Windows\StartMenu\Programs\Startup folder, and probably four or five other possibilities.

**Comment by drivin33** [comment]
I did try those but no luck. My cab has a button mapped to the escape key and pressing that button for a few seconds does return the focus to PinballY. I’ll keep trying different things until I figure it out. You have been very helpful - thanks again.

**Comment by drivin33** [comment]
Just to close the loop on this, I ended up installing AutoHotkey on the cab (which I was going to do anyway) and I created a script that simply presses the Escape key after PinballY loads. This returns focus to PinballY and solves the problem. One final question - What is the preferred method of loading PinVol at system startup? I now have a batch file in the Windows startup folder that loads PinballY, followed by the aforementioned AHK script. This work well, and I added a line to also run PinVol, but it will not run. Nor will PinVol run if I run it from the Task Scheduler at login. There are no errors or messages, it just doesn't run. The only way I'm able to get it to run at startup is to put a PinVol shortcut in the Windows startup folder along with my batch file. This works and am OK with it, but it would be a bit cleaner if I could run PinVol from the batch file.

**Comment by mjrgh** [comment]
Thanks for the update - glad you found a solution. I still don't know what's really going on, but AutoHotKey seems like a good workaround.

I'm not sure what's going wrong with the PinVol launch, other than that maybe it's somehow related to whatever's causing the focus issue. PinVol is a C# application based on an older Microsoft technology called Windows Forms, dating back to probably Windows 7 or 8, and that seems to have some known (but not well understood) problems with window initialization at program startup on modern Windows systems. I had to add a bunch of code to work around Windows Forms problems with launching minimized on Win 10, for example,  Maybe it's something related to that - whatever it is with your startup desktop context that's blocking focus from PinballY might also be messing with the fragile Windows Forms window initialization procedure. I'm guessing that the mysteriously disappearing process just means that the program is crashing during startup, which is what I'd expect if this is the explanation. So anyway, if we could get to the root of the PinballY focus issue, that might also coincidentally fix the PinVol launch problem, but I don't know what else to look at. So I guess I'd just stick with your workarounds for now.
